onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc

```diff
@@ -0,0 +1,242 @@
+  const Tensor* gamma = context->Input<Tensor>(5);
+  const Tensor* beta = context->Input<Tensor>(6);
+  const Tensor* mask = context->Input<Tensor>(7);  // optional. nullptr if not provided
+  const Tensor* word_embedding_scale_tensor = context->Input<Tensor>(8);
+  const Tensor* position_embedding_scale_tensor = context->Input<Tensor>(9);
+  const Tensor* segment_embedding_scale_tensor = context->Input<Tensor>(10);
+  const Tensor* gamma_scale_tensor = context->Input<Tensor>(11);
+  const Tensor* beta_scale_tensor = context->Input<Tensor>(12);
+  const Tensor* word_embedding_zero_point_tensor = context->Input<Tensor>(13);
+  const Tensor* position_embedding_zero_point_tensor = context->Input<Tensor>(14);
+  const Tensor* segment_embedding_zero_point_tensor = context->Input<Tensor>(15);
+  const Tensor* gamma_zero_point_tensor = context->Input<Tensor>(16);
+  const Tensor* beta_zero_point_tensor = context->Input<Tensor>(17);
+
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(word_embedding_scale_tensor),
+                    "Word embedding scale must be a scalar or 1D tensor of size 1");
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(position_embedding_scale_tensor),
+                    "Position embedding scale must be a scalar or 1D tensor of size 1");
+  if (segment_embedding != nullptr) {
+    ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(segment_embedding_scale_tensor),
+                      "Segment embedding scale must be a scalar or 1D tensor of size 1");
+  }
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(gamma_scale_tensor),
+                    "Gamma scale must be a scalar or 1D tensor of size 1");
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(beta_scale_tensor),
+                    "Beta scale must be a scalar or 1D tensor of size 1");
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(word_embedding_zero_point_tensor),
+                    "Word embedding zero point must be a scalar or 1D tensor of size 1");
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(position_embedding_zero_point_tensor),
+                    "Position embedding zero point must be a scalar or 1D tensor of size 1");
+  if (segment_embedding != nullptr) {
+    ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(segment_embedding_zero_point_tensor),
+                      "Segment embedding zero point must be a scalar or 1D tensor of size 1");
+  }
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(gamma_zero_point_tensor),
+                    "Gamma zero point must be a scalar or 1D tensor of size 1");
+  ORT_RETURN_IF_NOT(IsScalarOr1ElementVector(beta_zero_point_tensor),
+                    "Beta zero point must be a scalar or 1D tensor of size 1");
+
+  bool has_segment_embedding = segment_embedding != nullptr;
+
+  const int32_t* input_ids_data = input_ids->template Data<int32_t>();
+  const int32_t* segment_ids_data =
+      has_segment_embedding ? segment_ids->template Data<int32_t>() : nullptr;
+
+  const auto& input_dims = input_ids->Shape().GetDims();
```

Comment:
Do we restrict `input_ids` to be 2-dimensional? We probably should, given how it is used below.